system_zero/src/column_layout.rs

```diff
@@ -17,12 +17,131 @@ pub(crate) const COL_FRAME_PTR: usize = COL_INSTRUCTION_PTR + 1;
 /// Pointer to the tip of the current call's stack frame.
 pub(crate) const COL_STACK_PTR: usize = COL_FRAME_PTR + 1;
 
-//// PERMUTATION UNIT
+//// ARITHMETIC UNIT
+pub(crate) mod arithmetic {
+    //! Arithmetic unit.
+    const START_UNIT: usize = super::COL_STACK_PTR + 1;
+
+    pub(crate) const IS_ADD: usize = START_UNIT;
+    pub(crate) const IS_SUB: usize = IS_ADD + 1;
+    pub(crate) const IS_MUL: usize = IS_SUB + 1;
+    pub(crate) const IS_DIV: usize = IS_MUL + 1;
+
+    const START_SHARED_COLS: usize = IS_DIV + 1;
+
+    /// Within the arithmetic unit, there are shared columns which can be used by any arithmetic
+    /// circuit, depending on which one is active this cycle.
+    // Can be increased as needed as other operations are implemented.
+    const NUM_SHARED_COLS: usize = 3;
+
+    const fn shared_col(i: usize) -> usize {
+        debug_assert!(i < NUM_SHARED_COLS);
+        START_SHARED_COLS + i
+    }
+
+    /// The first value to be added; treated as an unsigned u32.
+    pub(crate) const COL_ADD_INPUT_1: usize = shared_col(0);
+    /// The second value to be added; treated as an unsigned u32.
+    pub(crate) const COL_ADD_INPUT_2: usize = shared_col(1);
+    /// The third value to be added; treated as an unsigned u32.
+    pub(crate) const COL_ADD_INPUT_3: usize = shared_col(2);
+
+    // Note: Addition outputs three 16-bit chunks, and since these values need to be range-checked
+    // anyway, we might as well use the range check unit's columns as our addition outputs. So the
+    // three proceeding columns are basically aliases, not columns owned by the arithmetic unit.
+    /// The first 16-bit chunk of the output, based on little-endian ordering.
+    pub(crate) const COL_ADD_OUTPUT_1: usize = super::range_check_16::col_rc_16_input(0);
+    /// The second 16-bit chunk of the output, based on little-endian ordering.
+    pub(crate) const COL_ADD_OUTPUT_2: usize = super::range_check_16::col_rc_16_input(1);
+    /// The third 16-bit chunk of the output, based on little-endian ordering.
+    pub(crate) const COL_ADD_OUTPUT_3: usize = super::range_check_16::col_rc_16_input(2);
+
+    pub(super) const END_UNIT: usize = START_UNIT + NUM_SHARED_COLS - 1;
+}
+
+pub(crate) mod logic {
+    //! Logic unit.
+    const START_UNIT: usize = super::arithmetic::END_UNIT + 1;
+    pub(super) const END_UNIT: usize = START_UNIT;
+}
+
+pub(crate) mod boolean {
+    //! Boolean unit. Contains columns whose values must be 0 or 1.
+
+    const START_UNIT: usize = super::logic::END_UNIT + 1;
+
+    const NUM_BITS: usize = 128;
+
+    pub const fn col_bit(index: usize) -> usize {
+        debug_assert!(index < NUM_BITS);
+        START_UNIT + index
+    }
+
+    pub(super) const END_UNIT: usize = START_UNIT + NUM_BITS - 1;
+}
```

Comment:
Oh forgot to explain that. It just holds columns that must be boolean. It's mainly intended for use by the "logic unit" (for splitting inputs so it can do bitwise ops on their bits). But it could also be used for other things, e.g. the arithmetic unit could use it for certain range checks (to reduce the number of table-based range checks we do).

Earlier we were talking about a "decomposition unit" which has input columns and bit columns, but I think we actually don't need dedicated input columns, e.g. for bitwise ops we can use a column owned by the "memory bus" as the input column.

Reply:
Oh cool, makes sense 👍